ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/processes/MachineNodeRenderStrategy.java

```diff
@@ -0,0 +1,158 @@
+import org.eclipse.che.ide.terminal.HasAddTerminalClickHandler;
+import org.eclipse.che.ide.ui.Tooltip;
+import org.eclipse.che.ide.util.dom.Elements;
+import org.vectomatic.dom.svg.ui.SVGImage;
+
+import static org.eclipse.che.api.core.model.workspace.WorkspaceStatus.RUNNING;
+import static org.eclipse.che.ide.ui.menu.PositionController.HorizontalAlign.MIDDLE;
+import static org.eclipse.che.ide.ui.menu.PositionController.VerticalAlign.BOTTOM;
+
+/**
+ * Strategy for rendering a machine node.
+ *
+ * @author Vlad Zhukovskyi
+ * @see ProcessTreeNodeRenderStrategy
+ * @see HasAddTerminalClickHandler
+ * @see HasPreviewSshClickHandler
+ * @since 5.10.0
+ */
+@Singleton
+public class MachineNodeRenderStrategy implements ProcessTreeNodeRenderStrategy, HasAddTerminalClickHandler, HasPreviewSshClickHandler {
+    private final MachineResources         resources;
+    private final CoreLocalizationConstant locale;
+    private final AppContext               appContext;
+    private final MachineMonitors          machineMonitors;
+
+    private AddTerminalClickHandler addTerminalClickHandler;
+    private PreviewSshClickHandler  previewSshClickHandler;
+
+    @Inject
+    public MachineNodeRenderStrategy(MachineResources resources,
+                                     CoreLocalizationConstant locale,
+                                     AppContext appContext,
+                                     MachineMonitors machineMonitors) {
+        this.resources = resources;
+        this.locale = locale;
+        this.appContext = appContext;
+        this.machineMonitors = machineMonitors;
+    }
+
+    @Override
+    public SpanElement renderSpanElementFor(ProcessTreeNode candidate) {
+        return createMachineElement(candidate);
+    }
+
+    private SpanElement createMachineElement(final ProcessTreeNode node) {
+        final MachineEntity machine = (MachineEntity)node.getData();
+        final String machineId = machine.getId();
+        final MachineConfig machineConfig = machine.getConfig();
+
+        SpanElement root = Elements.createSpanElement();
+
+        Element statusElement = Elements.createSpanElement(resources.getCss().machineStatus());
+        root.appendChild(statusElement);
+
+        if (node.isRunning()) {
+            statusElement.appendChild(Elements.createDivElement(resources.getCss().machineStatusRunning()));
+        } else {
+            statusElement.appendChild(Elements.createDivElement(resources.getCss().machineStatusPausedLeft()));
+            statusElement.appendChild(Elements.createDivElement(resources.getCss().machineStatusPausedRight()));
+        }
+
+        Tooltip.create(statusElement, BOTTOM, MIDDLE, locale.viewMachineRunningTooltip());
+
+        Workspace workspace = appContext.getWorkspace();
+        if (workspace != null && RUNNING == workspace.getStatus() && node.hasTerminalAgent()) {
+            SpanElement newTerminalButton = Elements.createSpanElement(resources.getCss().newTerminalButton());
```

Comment:
I think, we should create separate method for creating new terminal element "createNewTerminalElement" and use it here.